README.markdown
JQuery Blank
============

Credits
--------

 * Jeff Hui - jeffhui.net
 * Ashish - ashish@setfive.com
 * Sherri Alexander - sherri-alexander.com

About
--------

JQuery Blank is a simple way to create default values for text fields and textareas
which hints to the user what should be filled into the field.
 
A good example is [mobile me's][1] login text input fields. You can also see a [demo here][2].

When the user focuses on the textbox, the search text is instantly removed and the
user can then type something. If the user leaves the textbox and if the textbox is
 still empty, the default text is restored.

Simplest usage:

`$(myelement).blank('Search');`

Documentation
=============

If you use [HTML5 placeholder][4] attribute for all your input elements, you can
simply do:

`$.blankify();`

in the onready event. Only one other function is added:

`jQuery.fn.blank([default_text[, options]]);`

* `default_text`: The default text to display for the given element.
* `options`: Additional parameters to pass: (object)
    * `defaultCSS`: The css style when the text input field is empty. This can be either a object (hash) or a function call that returns an object. The function contains the element as this and as the first parameter. The default value is {color: '#999'}.
    * `normalCSS`: The css style when the text has entered user input. This can also be either an object or a function that returns one. The parameters passed to the function are the element, the object of defaultCSS, and the object of d. The element is also passed as this. The default is a function that merges both defaultCSS and defaultFocusCSS for style attributes and computates the style values for the element when it is called.
    * `defaultFocusCSS`: The css style when the text input field has user focus, but is still blank (user didn't type anything yet). This is used to for a similar effect like Apple's me.com login box. This style only applies when defaultUntilType is set to true. Defaults to {color: '#bbb'}.
    * `defaultUntilType`: If enabled, when the user focuses on the text input, the defaultFocusCSS style is applied and the default text is display until the user types something. Once the user types something, the default text is removed and the user can type input.

If no parameters are specified, then the function returns whether or not the text 
field is blank. True if the input text is blank, false if not.

Example usage:

    $(function(){ $('#myform').blank('Required'); });
	$('#myform').submit(function(){
	  if($('#myinput').blank()){
		alert('Please fill out the input box!');
		return false;
	  }
	  return true;
	});
	
[4]: http://davidwalsh.name/html5-placeholder

Events
==========

Since version 1.4, two events exist for jquery.blank elements. They can be hooked using `$.bind('<trigger-name>', function(){});`

* `blank`: This is called when the text field is set to blank. For *defaultUntilType* enabled fields, this is on keyup, and onblur for other text fields.
* `fill`: This is called when the text field is filled by user input. For *defaultUntilType* enabled fields, this is on keydown, and onblur for other text fields.


License
=========

[MIT License][3]

 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:
 
 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.


[1]: http://me.com/ "Apple's Mobile Me"
[2]: http://demo.jeffhui.net/jquery-blank/ "Example of JQuery.Blank"
[3]: http://creativecommons.org/licenses/MIT/



changelog
Changelog:

TODO:
 o Refactor code

--- Version 1.5
 o Fixed $.fn.blank() returning True if the user submitted the form without leaving
   the input field. (reported by Ashish Datta)
 o Fixed $.fn.blank() from absorbing tabbing through input fields when defaultUntilType
   is true.
 o Support for HTML5 placeholder attribute by:
   o jQuery.support.input_placeholder is true if the HTML5 placeholder attribute is
     supported by the current browser
   o Use $.blankify() to automatically call el.blank(el.attr('placeholder')) on all
     input/textareas with the placeholder attributes.

--- Version 1.4
 o Rudimentary fix for not allowing home/end the defaultText (with defaultUntilType)
 o Added password field support (by Ashish Datta)
 o Fixed blank value not set on page reload (by Ashish Datta)
 o Fixed blank value not set on form reset
 o Fixed blank value being selected when tab-focused under defaultUntilType.
 o Code formatting fixes
 o Added two events: 'blank' and 'fill', which are used for password fields

--- Version 1.3
 o Refactored code
 o Changed onsubmit behavior for blank text to be empty string
   (credits: Ashish - ashish@setfive.com)

--- Version 1.2
 o Created changelog
 o Fixed indexof bug for IE (reported by Sherri Alexander)


jquery.blank.js
/** jquery.blank ver. 1.5 - JQuery Plugin
 * =============
 * Credits:
 *  o Jeff Hui - jeffhui.net
 *  o Ashish - ashish@setfive.com
 *  o Sherri Alexander - sherri-alexander.com
 * =============
 * Copyright (c) 2009 Jeff Hui <contrib@jeffhui.net>
 * 
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * =============
 * vim: tabstop=4;shiftwidth=4;noexpandtab
 */
(function($){

// keys that we ignore for keypress (for defaultUntilType)
var ignoreList = [
	8,  // backspace
	9,  // tab
	16, // shift
	17, // alt
	18, // pause/break
	20, // capslock
	27, // esc
	33, // pgup
	34, // pgdown
	35, // end
	36, // home
	91, // left winkey
	92, // right winkey
	144,// numlock
	145 // scroll lock
];
// keys that do a movement in the textbox that we should suppress
// we can assume we're at the left-most side
var moveList = [
	34, 35 // pgdown, end
];

var fill_triggername = 'fill',
	blank_triggername = 'blank';

// Gets the value of func, calling it if its a function. el nas param
// are values passed into it.
function call(func, el, param){
	if($.isFunction(func))
		return (param != undefined) ? func.call(el, el, param) : func.call(el, el);
	else
		return func;
}

// moves the text caret for the element to beginning of the textfield.
// accepts the raw element (not the jQuery object).
// Optionally accepts an optional position to set the caret to.
function reset_caret(e, pos){
	if(pos === undefined)
		pos = 0;
	if(e.selectionStart !== undefined){ // W3C compliant browsers
		e.selectionStart = pos;
		e.selectionEnd = pos;
	} else if(e.createTextRange){ // IE-browsers
		var tr = e.createTextRange();
		tr.collapse(true);
		tr.moveStart('character', pos);
		tr.moveEnd('character', pos);
		tr.select();
	}
}

$.support.input_placeholder = 'placeholder' in document.createElement('input');


var supported_elements = 'input[type=text], input[type=password], textarea';
$.blankify = function(){
	if(!$.support.input_placeholder){
		$(supported_elements).each(function(){
			var el = $(this);
			el.blank(el.attr('placeholder'));
		});
	}
}

////// main function //////
// assume: the user knows this only should be used only with text-typing elements
$.fn.blank = function(text, options){

	// when we use $(el).blank() # returns a true/false if the input is actually blank
	if(text === undefined){
		var d = this.data('blank');
		return d !== undefined;
	}

	// gather options
	var opt = {
		defaultCSS: {color: '#999'},
		normalCSS: function(el, defCSS, defFocusCSS){
			// here we just try to get default css styles from the element
			// for all key values of defCSS
			var css = $.extend({}, defCSS, defFocusCSS);
			var e = $(this);
			$.each(defCSS, function(i, v){
				css[i] = $.curCSS(e.get(0), i);
			});
			return css;
		},
		defaultFocusCSS: {color: '#bbb'},
		defaultUntilType: false
	};
	$.extend(opt, options);

	// function to work on each element we got
	return this.each(function(){
		if(!$(this).is(supported_elements))
		  return true; // continue

		// css calls are in here, instead of out this call because
		// option could be a function; which requires the element parameter
		var defCSS = call(opt.defaultCSS, this),
			defFocusCSS = call(opt.defaultFocusCSS, this, defCSS),
			normCSS = call(opt.normalCSS, this, defCSS, defFocusCSS),
			el = $(this),
			elBlankForm; // used by password fields only

		// If the current text is blank text, wipe it since it might be a CTRL+R refresh
		if(el.val() === text)
			el.val('');

		// Password field support:
		// We can't change the type=password attribute, so we have to use a new textfield
		// that is swapped when we need to show default text.
		if(el.is(':password')){
			elBlankForm = el.after('<input type="text" class="jqblank_field" />').next();
			elBlankForm.blank(text, options);
			el.hide();

			if(opt.defaultUntilType){
				elBlankForm.bind('fill', function(){
					elBlankForm.hide();
					el.val(elBlankForm.val()).show().focus();
					reset_caret(el, 1);
				});
				el.keyup(function(evt){
					if(el.val() === ''){
						el.hide();
						elBlankForm.val('').show().focus().trigger('keyup', [evt]);
						reset_caret(elBlankForm, 1);
					}
				});
			} else {
				elBlankForm.focus(function(){
					elBlankForm.hide();
					el.show().focus();
				});
				el.blur(function(){
					if($.trim(el.val()) === ''){
						el.hide();
						elBlankForm.show().blur();
					}
				});
			}
		}

		// if the elements are inside a form, revert to blank text
		el.closest('form').submit(function(){
			if(el.blank())
				el.val('');
			return true;
		}).bind('reset', function(){
			setTimeout(function(){
				if(el.is(':password')){
					if(opt.defaultUntilType)
						el.hide().next().show().blur().keyup();
					else
						el.focus().val('').keyup().blur();
				} else
					el.keyup().blur();
			}, 1);
		});

		// don't continue if we're a password field
		if(elBlankForm)
			return true; // continue
		
		if(!opt.defaultUntilType){

			el.blur(function(){
				var isBlank = el.blank();
				if($.trim(el.val()) === ''){
					el.val(text).css(defCSS).data('blank', 1);
					if(!isBlank)
						el.trigger(blank_triggername);
				} else {
					el.removeData('blank');
					if(isBlank)
						el.trigger(fill_triggername);
				}
				
			}).focus(function(){
				
				if(el.blank())
					el.val('').css(normCSS);
					
			// we need keyhooks incase the user presses enter.
			}).keydown(function(evt){

				if(el.blank() && evt.keyCode !== 13 && $.inArray(evt.keyCode, ignoreList) === -1)
					el.removeData('blank').trigger(fill_triggername);

			}).keyup(function(){

				if(!el.blank() && el.val() === '')
					el.data('blank', 1).trigger(blank_triggername);

			}).blur();

		} else { // opt.defaultUntilType

			el.data('blank', 1);
			el.blur(function(){
				if(el.blank()){
					el.css(defCSS).val(text);
				}
			}).mouseup(function(){

				if(!el.blank()) return true;
				reset_caret(el.get(0));

			}).focus(function(){

				if(el.blank()){
					el.css(defFocusCSS);
					// disable text selection on tab-focus
					setTimeout(function(){reset_caret(el.get(0));}, 1);
				}

			}).keydown(function(evt){

				if($.inArray(evt.keyCode, ignoreList) !== -1){
					if(el.blank() && $.inArray(evt.keyCode, moveList) !== -1)
						setTimeout(function(){reset_caret(el.get(0))}, 1);
					return true;
				}
				var v = el.val();
				if(el.blank()){
					el.val(v.substring(0, v.length-text.length))
					.removeData('blank')
					.css(normCSS)
					.trigger(fill_triggername);
				}

			}).keyup(function(){

				if(!el.blank() && el.val() === '')
					el.data('blank', 1).css(defFocusCSS).trigger(blank_triggername);

			}).blur();
		} // if-else
	}); // return this.each(function(){})

}; // $.fn.blank = function(){}

})(jQuery);
